hepcrawl/spiders/oaipmh_spider.py

```diff
@@ -0,0 +1,100 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of hepcrawl.
+# Copyright (C) 2015, 2016, 2017 CERN.
+#
+# hepcrawl is a free software; you can redistribute it and/or modify it
+# under the terms of the Revised BSD License; see LICENSE file for
+# more details.
+
+"""Generic spider for OAI-PMH servers."""
+
+import logging
+import sickle
+from datetime import datetime
+
+from sickle import Sickle
+from sickle.models import Record
+from sickle.oaiexceptions import NoRecordsMatch
+
+from scrapy.http import Request
+from scrapy.spiders import Spider
+
+logger = logging.getLogger(__name__)
+
+class OAIPMHSpider(Spider):
+    """
+    Implements a spider for the OAI-PMH protocol by using the Python sickle library.
+
+    In case of successful harvest (OAI-PMH crawling) the spider will remember the initial starting
+    date and will use it as `from_date` argument on the next harvest.
+    """
+    name = 'OAI-PMH'
+    state = {}
+
+    def __init__(self, url, metadata_prefix='marcxml', set=None, alias=None, from_date=None, until_date=None, granularity='YYYY-MM-DD', record_class=Record, *args, **kwargs):
+        super(OAIPMHSpider, self).__init__(*args, **kwargs)
+        self.url = url
+        self.metadata_prefix = metadata_prefix
+        self.set = set
+        self.granularity = granularity
+        self.alias = alias or self._make_alias()
+        self.from_date = from_date
+        logger.info("Current state:{}".format(self.state))
+        self.until_date = until_date
+        self.record_class = record_class
+
+    def start_requests(self):
+        self.from_date = self.from_date or self.state.get(self.alias)
+        logger.info("Current state 2:{}".format(self.state))
+        logger.info("Starting harvesting of {url} with set={set} and metadataPrefix={metadata_prefix}, from={from_date}, until={until_date}".format(
+            url=self.url,
+            set=self.set,
+            metadata_prefix=self.metadata_prefix,
+            from_date=self.from_date,
+            until_date=self.until_date
+        ))
+        now = datetime.utcnow()
+        request = Request('oaipmh+{}'.format(self.url), self.parse)
+        yield request
+        self.state[self.alias] = self._format_date(now)
```

Comment:
@david-caro it does support state management, provided the spider is launched with the appropriate flag.

Reply:
The state is not persisted, it only exists in this instance of the spider (it should be using also the `StatefulSpider`). We have to find a way to properly persist it. The one we were using before (by enabling the job dir https://doc.scrapy.org/en/latest/topics/jobs.html#keeping-persistent-state-between-batches), does not work for parallel runs on scrapyd.

Reply:
cannot you adapt the crawl once thing to work here?

Reply:
crawl-once uses a local db to store key-values, maybe we would be able to use it for this, though I'm not sure if it's abusing too much.